### PR TITLE
Failure-safe Admin Logging

### PR DESCRIPTION
**Pull request by pandabear15**
Implemented log-before-write. If writing to the QLDB fails, the endpoint returns a 500 error and does not apply any changes requested from the user, if applicable. Most of the changes in the services are procedural ones, but EncryptionKeyService.java required a fair amount of reorganizing to implement. No test cases were changed, so the end behavior should still be the same assuming writing to QLDB is successful.

Implemented itemType and itemKey as top-level fields for the purposes of indexing on the QLDB. Also implemented indexing on the QLDB, which is necessary to keep infra costs manageable at large database sizes.

Added a new QLDB entry for the null item. This entry is written to in the case where a client writes (correctly) to an endpoint expecting possibly multiple items in the S3 to be modified, but no item is actually modified. Currently, only a single service (EncryptionKeyService.java) writes to the item in the QLDB.

### CODE REVIEW

File: src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java [hunk 1]
```diff
@@ -11,47 +11,79 @@
...
+                            txn.abort();
+                            break;
+                        }
+                        QLDBAuditModel qldbModel = (QLDBAuditModel) model;
+                        JsonObject jsonObject = qldbModel.writeToJson();
+                        String query;
+                        List<IonValue> sanitizedInputs = new ArrayList<>();
+                        if(qldbModel.actionTaken == Actions.CREATE){
+                            query = "INSERT INTO " + logTable + " VALUE ?";
+                            JsonObject wrapped = new JsonObject().put("data", jsonObject);
```
Comment (haitaottd) [inline src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java:57]: are you sure this puts itemType and itemKey at top level? Seems to me they are nested under `data` field

File: src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java [hunk 2]
```diff
@@ -11,47 +11,79 @@
...
-                sanitizedInputs.add(ionSys.newLoader().load(jsonObject.toString()).get(0));
-                query.append(" WHERE t.data.itemType = ? AND t.data.itemKey = ?");
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemType")));
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemKey")));
-
-                Result r = txn.execute(query.toString(), sanitizedInputs);
-                if (!r.iterator().hasNext()) {
-                    logger.warn("Malformed audit log input: no log written to QLDB");
+    public boolean writeLogs(Collection<AuditModel> models) {
+        AtomicBoolean successfulLog = new AtomicBoolean(true);
```
Comment (haitaottd) [inline src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java:40]: I don't see multi-threading here, why use atomic?

File: src/main/java/com/uid2/admin/audit/QLDBInit.java [hunk 3]
```diff
@@ -1,350 +1,134 @@
...
-            throw new RuntimeException("AWS configuration not set up");
+            LOGGER.warn("qldb not initialized");
         }
     }
 
-    private static boolean isTableSetup() {
+    private static boolean hasTableBeenCreated() {
         try {
-            final IonStruct[] count = new IonStruct[1];
+            AtomicBoolean hasTableBeenCreated = new AtomicBoolean(false);
```
Comment (haitaottd) [inline src/main/java/com/uid2/admin/audit/QLDBInit.java:50]: don't think atomic is necessary

File: src/main/java/com/uid2/admin/audit/QLDBInit.java [hunk 4]
```diff
@@ -1,350 +1,134 @@
...
-            jo.put("max_last_updated", Arrays.stream(lastSnapshot.getAllRotatingSalts())
-                    .map(SaltEntry::getLastUpdated)
-                    .max(Long::compare).orElse(null));
-
-            Collection<OperationModel> newModels = new HashSet<>();
-            newModels.add(new OperationModel(Type.SALT, "singleton", null,
-                    DigestUtils.sha256Hex(jo.toString()), null));
-            models.add(newModels);
+            qldbDriver.execute(txn -> {
+                txn.execute("CREATE INDEX ON " + qldbTableName + "(itemType)");
```
Comment (haitaottd) [inline src/main/java/com/uid2/admin/audit/QLDBInit.java:90]: why don't create one index on two fields

File: src/main/java/com/uid2/admin/audit/QLDBInit.java [hunk 5]
```diff
@@ -1,350 +1,134 @@
...
-            e.printStackTrace();
-        }
+    private static void insertIntoQLDB(OperationModel model) { //populates qldb
+        QLDBAuditModel auditModel = new QLDBAuditModel(model.itemType, model.itemKey, model.actionTaken, null,
+                null, null, -1, model.itemHash, model.summary);
+        qldbDriver.execute(txn -> {
+            JsonObject jsonObject = new JsonObject();
+            jsonObject.put("data", auditModel.writeToJson());
+            txn.execute("INSERT INTO " + qldbTableName + " VALUE ?",
+                    ionSys.newLoader().load(auditModel.writeToJson().toString()).get(0));
```
Comment (haitaottd) [inline src/main/java/com/uid2/admin/audit/QLDBInit.java:105]: again, does this put itemType and itemKey at top level

File: src/main/java/com/uid2/admin/audit/QLDBInit.java (hunk 4 above)
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBInit.java:90]: QLDB does not support an index on multiple fields

File: src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java (hunk 1 above)
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java:57]: Thank you for the catch, forgot to change this

File: src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java [hunk 6]
```diff
@@ -11,47 +11,79 @@
...
+        try {
+            if (qldbLogging) {
```
[the remark this answers is not in the document]
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java:42]: will do

File: src/main/java/com/uid2/admin/audit/QLDBAuditModel.java [hunk 7]
```diff
@@ -62,7 +62,14 @@ public QLDBAuditModel(Type itemType, String itemKey, Actions actionTaken, String
     @Override
     public JsonObject writeToJson() {
         Gson gson = new GsonBuilder().serializeNulls().create();
-        return new JsonObject(gson.toJson(this));
+        JsonObject jo = new JsonObject(gson.toJson(this));
+        jo.remove("itemType");
```
[the remark this answers is not in the document]
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBAuditModel.java:66]: will do

File: src/main/java/com/uid2/admin/audit/QLDBInit.java (hunk 5 above)
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBInit.java:105]: fixed

File: src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java (hunk 2 above)
Comment (pandabear15) [reply src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java:40]: Atomic here is just serving as a container for a boolean as anything referenced in the lambda expression must be "effectively final" or a compile error is thrown

**Comment by aulme** [comment]
A different strategy for admin audit was selected for implemetnation.

**Pull request closed without merging**